### ISSUE TITLE
TypeError: a bytes-like object is required, not 'str' in line 74 yowsup/structs/protocoltreenode.py when sync contact

### ISSUE THREAD

**Issue by Sharissa90** [tgalal/yowsup#2796]
**Describe the bug**

~~Trying new commit including this commit https://github.com/tgalal/yowsup/commit/031d1e9302a8a8779625286f8ac1ff8a59e3f892 from master branch and get TypeError: a bytes-like object is required, not 'str' in yowsup/structs/protocoltreenode.py~~

**UPDATE FIX**
```
I purpose to change line 74 from:
                data = "0x%s" % binascii.hexlify(data.encode()).decode()
                out = "%s\n%s%s%s" % (out, self._STR_INDENT, data, postfix)
to:
                if isinstance(data, bytes):
                    data = "0x%s" % binascii.hexlify(data).decode()
                else:
                    data = "0x%s" % binascii.hexlify(data.encode()).decode()
                out = "%s\n%s%s%s" % (out, self._STR_INDENT, data, postfix)
```
binascii.hexlify need bytes not string sometime variable data is string and sometimes is bytes, so to fix this issue we must check if data is bytes or string

```
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/noise/workers/handshake.py", line 42, in run
    self._protocol.start(self._stream, self._client_config, self._s, self._rs)
  File "/usr/local/lib/python3.5/dist-packages/consonance/protocol.py", line 88, in start
    self._machine.finish()
  File "/usr/local/lib/python3.5/dist-packages/transitions/core.py", line 383, in trigger
    return self.machine._process(func)
  File "/usr/local/lib/python3.5/dist-packages/transitions/core.py", line 1047, in _process
    return trigger()
  File "/usr/local/lib/python3.5/dist-packages/transitions/core.py", line 399, in _trigger
    return self._process(event_data)
  File "/usr/local/lib/python3.5/dist-packages/transitions/core.py", line 409, in _process
    if trans.execute(event_data):
  File "/usr/local/lib/python3.5/dist-packages/transitions/core.py", line 266, in execute
    machine.callback(func, event_data)
  File "/usr/local/lib/python3.5/dist-packages/transitions/core.py", line 1005, in callback
    func(*event_data.args, **event_data.kwargs)
  File "/usr/local/lib/python3.5/dist-packages/consonance/protocol.py", line 58, in _trigger_state_callback
    self._protocol_state_callbacks(self._machine.state)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/noise/layer.py", line 139, in _on_protocol_state_changed
    self._flush_incoming_buffer()
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/noise/layer.py", line 160, in _flush_incoming_buffer
    self.toUpper(self._wa_noiseprotocol.receive())
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 79, in toUpper
    self.__upper.receive(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/coder/layer.py", line 21, in receive
    self.toUpper(node)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 79, in toUpper
    self.__upper.receive(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/logger/layer.py", line 14, in receive
    self.toUpper(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 79, in toUpper
    self.__upper.receive(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/axolotl/layer_control.py", line 35, in receive
    self.toUpper(protocolTreeNode)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 79, in toUpper
    self.__upper.receive(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 194, in receive
    s.receive(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/axolotl/layer_receive.py", line 35, in receive
    self.toUpper(protocolTreeNode)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 79, in toUpper
    self.__upper.receive(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 194, in receive
    s.receive(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 130, in receive
    recv(node)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/auth/layer_authentication.py", line 53, in handleSuccess
    self.broadcastEvent(successEvent)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 202, in subBroadcastEvent
    self.broadcastEvent(yowLayerEvent)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 103, in broadcastEvent
    self.__lower.broadcastEvent(yowLayerEvent)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 98, in broadcastEvent
    if self.__lower and not self.__lower.onEvent(yowLayerEvent):
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 109, in onEvent
    return self.event_callbacks[eventName](yowLayerEvent)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/axolotl/layer_control.py", line 68, in onAuthed
    self._unsent_prekeys[:], reboot_connection=True
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/axolotl/layer_control.py", line 109, in flush_keys
    self._sendIq(setKeysIq, onResult, self.onSentKeysError)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 151, in _sendIq
    self.toLower(iqEntity.toProtocolTreeNode())
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/__init__.py", line 84, in toLower
    self.__lower.send(data)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/layers/logger/layer.py", line 8, in send
    logger.debug("tx:\n%s" % ldata)
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/structs/protocoltreenode.py", line 57, in __str__
    children = "\n".join(map(str, self.children))
  File "/usr/local/lib/python3.5/dist-packages/yowsup-3.2.3-py3.5.egg/yowsup/structs/protocoltreenode.py", line 74, in __str__
    data = "0x%s" % binascii.hexlify(data).decode()
TypeError: a bytes-like object is required, not 'str'
```

**Config file**

Not Related to Config

**Versions**

```
Python 3.5.3
---------------
Name: consonance
Version: 0.1.3.post1
Name: dissononce
Version: 0.34.3
Name: python-axolotl
Version: 0.2.2
Name: cryptography
Version: 2.6.1
```

**To Reproduce**

~~Using Latest Branch Master and have using this commit https://github.com/tgalal/yowsup/commit/031d1e9302a8a8779625286f8ac1ff8a59e3f892~~

**Expected behavior**
~~stable as previous update~~

**OS (please complete the following information):**
Debian GNU/Linux 9

**Additional context**
No

**Comment by tgalal** [comment]
I'm slightly confusing, did it turn out that  031d1e9 did fix the issue for you or issue still exists?

**Comment by Sharissa90** [comment]
@tgalal after I post this issue, I notice you made new commit https://github.com/tgalal/yowsup/commit/031d1e9302a8a8779625286f8ac1ff8a59e3f892 so I implemented your commit but not work, I still get TypeError: a bytes-like object is required, not 'str' so I tried to find out what happen with variable  data and I find variable data sometimes have string type and sometimes its bytes. so I made little change and test it and its work in my case.. you can test with print(type(data)) before data = "0x%s" % binascii.hexlify(data).decode() and it will show str and byte.

**Comment by tgalal** [comment]
that's true i'm aware of this and working on a different kind of fix  to make it consistent. My question however is whether the mentioned commit (031d1e9) does fix the exact error that appeas in logs you posted, that is, when yowsup sends a SetKeysIq which is what the commit addressed. If yes please close the issue, or update with logs of a different crash log associated to same issue.

**Comment by Sharissa90** [comment]
the commit https://github.com/tgalal/yowsup/commit/031d1e9302a8a8779625286f8ac1ff8a59e3f892 doesn't fix the issue if I using my app layer, but the weird is if I using yowsup-cli demos the error doesn't show up, I think I must investigate my current layer app, and I close this issue.. Thank you @tgalal

**Comment by Sharissa90** [comment]
I have found the cause of my issue, When I sync contact
```
        contacts = self.getProp(self.__class__.PROP_CONTACTS, [])
        contact_entity = GetSyncIqProtocolEntity(contacts)
        self._sendIq(contact_entity, self.on_sync_result, self.on_sync_error)
        self.toLower(ListGroupsIqProtocolEntity())
```
the issue is show up.. But If I don't sync contact its running smoothly. I think I must set my contact number to bytes. Or you have another idea @tgalal ?

**Comment by tgalal** [comment]
@Sharissa90 with the fixed I've just pushed (69c5211) you should revert your contacts dict to have normal str keys, so

```
contacts = {
     "99xxxxxxxxx":"friend1",
     "99xxxxxxxxx":"friend2"
}
```
and not
```
contacts = {
     b"99xxxxxxxxx":"friend1",
     b"99xxxxxxxxx":"friend2"
}
```

And the error happened in your app and not yowsup-cli because the error is triggered on logging ProtocolTreeNode objects. You must have log level Debug set in your application. In yowsup-cli if you pass `--debug` you would have seen the same error.
